src/compiler/checker.ts

```diff
@@ -6925,7 +6966,11 @@ namespace ts {
             checkCollisionWithCapturedThisVariable(node, node);
             checkBlockScopedBindingCapturedInLoop(node, symbol);
 
-            return getNarrowedTypeOfSymbol(getExportSymbolOfValueSymbolIfExported(symbol), node);
+            const type = getNarrowedTypeOfSymbol(getExportSymbolOfValueSymbolIfExported(symbol), node);
+            if (type === undefinedType || type == nullType) {
+                return (contextualMapper || identityMapper)(type);

```

Comment:
The `contextualMapper` parameter doesn't indicate whether you have a contextual type. You need to actually get the contextual type using `getContextualType`.